Document internal Prometheus server architecture

juliusv opened the pull request:
This is a WIP to address https://github.com/prometheus/prometheus/issues/4219

At this point I am only interested in feedback on the overall approach and whether we want to include this document in the documentation site on prometheus.io/docs or whether it should live elsewhere. It *is* mostly for developers wanting to understand internals and get started with contributing, but it could also be interesting for users to understand the Prometheus server's internals better.

I will still add more details to each section and of course the architecture diagram has to be done properly and all.

brian-brazil commented:
I'd say we put it under the "documentation" directory for now, and link that from the contributing guide.

henridf commented:
@juliusv This is a great start! 

One complementary set of content I can see would be to explain the journey of a given request through the read and write paths. For example for read, consider an incoming request to the `query` endpoint, and describe the components it goes through and the handling decisions made along the way. Similarly for scrapes. I think it could help tie things together - if the current content is a log, this would be a trace :)

juliusv commented:
@henridf Great idea, totally. Actually I wrote the "Life of a query" doc you propose once in the very early days (2014) in the wiki, but deleted it later when we got rid of the wiki (and probably it was outdated already). I cloned the wiki, dug around in its history, and copied the old "Life of a query" doc into this gist for anyone curious about it (of course today the flow would be somewhat different): https://gist.github.com/juliusv/0c9588f95ed6940412baae98a0dae1f8

henridf commented:
@juliusv oh nice. i can draft up the 2018 version if that's of any use, but probably best after the current piece starts to converge.

juliusv commented:
I moved this to the `documentation` directory, added a lot more details about the scrape manager internals and removed the [WIP]. The diagram at the top is still hand drawn and should be re-done digitally, but I think it's a good idea to get this out there quick and I'll do the image properly in a bit.

krasi-georgiev commented:
looks good , very good start. 
Like how tsdb is in the middle of everything, never thought of it this way.

brian-brazil commented on documentation/internal_architecture.md:
```diff
@@ -0,0 +1,129 @@
+
+The overall Prometheus server architecture is shown in this diagram:
+
+![Prometheus server architecture](https://i.imgur.com/FTeb80I.jpg)
```
Can we host these ourselves?

juliusv replied:
Oh sure, will do.

brian-brazil commented on documentation/internal_architecture.md:
```diff
@@ -0,0 +1,129 @@
+
+### Target scrapes
+
+Finally, a scrape loop periodically scrapes its targets over HTTP and tries to decode the received HTTP responses according to the [Prometheus text-based metrics exposition format](https://prometheus.io/docs/instrumenting/exposition_formats/). It then applies [metric relabeling configurations](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#%3Cmetric_relabel_configs%3E)  ...
```
I'd mention timeouts here, a scrape shouldn't take more than the interval. We get that question once every month or two.

brian-brazil commented on documentation/internal_architecture.md:
```diff
@@ -0,0 +1,129 @@
+
+### Local storage
+
+Prometheus's local on-disk time series database is a [light-weight wrapper](https://github.com/prometheus/prometheus/blob/v2.3.1/storage/tsdb/tsdb.go#L102-L106) around [`github.com/prometheus/tsdb.DB`](https://github.com/prometheus/tsdb/blob/master/db.go#L92-L117). The wrapper makes only minor interface adjustments for use of the TSDB in the context of the Prometheus server and implements the [`s ...
```
Remove TODO?

brian-brazil commented on documentation/internal_architecture.md:
```diff
@@ -0,0 +1,129 @@
+
+### Fanout storage
+
+The [fanout storage](https://github.com/prometheus/prometheus/blob/v2.3.1/storage/fanout.go#L27-L32) is a [`storage.Storage`](https://github.com/prometheus/prometheus/blob/v2.3.1/storage/interface.go#L31-L44) implementation that proxies and abstracts away the details of the underlying local and remote storage subsystems for use by other components. For reads, it merges query results from local an ...
```
Should probably mention here that it's planned not to always talk to remote read for rules.

brian-brazil commented on documentation/internal_architecture.md:
```diff
@@ -0,0 +1,129 @@
+
+## Web UI and API
+
+Prometheus serves its web UI and API on port `9090` by default. The web UI is available at `/` and serves a human-usable interface for running expression queries, inspecting active alerts, or getting other insight into the status of the Prometheus server. The web API is served under `/api/v1` and allows programmatical [querying, metadata, and server status inspection](https://prometheus.io/docs/p ...
```
Mention console templates?

juliusv commented:
@brian-brazil Comments addressed.

krasi-georgiev commented:
https://github.com/prometheus/prometheus/blob/master/documentation/internal_architecture.md
why not use the same tools to build the same graph as https://camo.githubusercontent.com/7cc17b981938e40974542fbfa9c34172fd92eccd/68747470733a2f2f63646e2e7261776769742e636f6d2f70726f6d6574686575732f70726f6d6574686575732f633334323537643036396336333036383564613335626365663038343633326666643564363230392f646f63756d656e746174696f6e2f696d616765732f6172636869746563747572652e737667

juliusv commented:
@krasi-georgiev That's my plan, actually was going to do this today :)

juliusv commented:
@krasi-georgiev Also: Y U NO RESPECT MY ART!!!!1 ;)

krasi-georgiev commented:
:stuck_out_tongue_winking_eye:
